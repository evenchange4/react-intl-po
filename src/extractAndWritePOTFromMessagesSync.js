/* eslint-disable no-console */
import fs from 'fs';
import chalk from 'chalk';
import flowRight from 'lodash/flowRight';
import readAllMessageAsObjectSync from './readAllMessageAsObjectSync';
import potFormater from './potFormater';

const customKeyMapper = (message, messageKey, filename) => ({
  [message[messageKey]]: [{ ...message, filename }],
});

const customKeyMapperFactory = (messageKey = 'defaultMessage') =>
  (message, filename) => customKeyMapper(message, messageKey, filename);

function extractAndWritePOTFromMessagesSync(srcPatterns, { messageKey, output }) {
  const mapper = messageKey ? customKeyMapperFactory(messageKey) : undefined;

  const result = flowRight(
    potFormater,                // 2. return formated string
    readAllMessageAsObjectSync, // 1. return messages object
  )(srcPatterns, mapper);

  fs.writeFileSync(output, result);
  console.log(chalk.green(`> [react-intl-po] write file -> ${output} ✔️\n`));
}

export default extractAndWritePOTFromMessagesSync;
