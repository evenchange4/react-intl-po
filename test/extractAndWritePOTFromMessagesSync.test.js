import fs from 'fs';
import extractAndWritePOTFromMessagesSync from '../src/extractAndWritePOTFromMessagesSync';

it('should return a function', () => {
  expect(typeof extractAndWritePOTFromMessagesSync).toBe('function');
});

it('should return messages object with default mapper', () => {
  const output = './test/temp/extract-defaultMessage.pot';
  const headerOptions = {
    potCreationDate: new Date(Date.UTC(2017, 1, 1, 11, 23, 12)),
  };

  extractAndWritePOTFromMessagesSync('./test/messages/**/*.json', {
    output,
    headerOptions,
  });
  expect(fs.readFileSync(output, 'utf8')).toMatchSnapshot();
});

it('should return messages object with custom message key mapper', () => {
  const output = './test/temp/extract-id.pot';
  const headerOptions = {
    potCreationDate: new Date(Date.UTC(2017, 1, 1, 11, 23, 12)),
  };

  extractAndWritePOTFromMessagesSync('./test/messages/**/*.json', {
    messageKey: 'id',
    output,
    headerOptions,
  });
  expect(fs.readFileSync(output, 'utf8')).toMatchSnapshot();
});

it('should return messages object with messageContext', () => {
  const output = './test/temp/extract-messageContext.pot';
  const headerOptions = {
    potCreationDate: new Date(Date.UTC(2017, 1, 1, 11, 23, 12)),
  };

  extractAndWritePOTFromMessagesSync('./test/messages/**/*.json', {
    output,
    headerOptions,
    messageContext: 'id',
  });
  expect(fs.readFileSync(output, 'utf8')).toMatchSnapshot();
});
