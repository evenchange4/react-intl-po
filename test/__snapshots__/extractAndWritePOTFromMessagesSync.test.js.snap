// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return messages object with custom message key mapper 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2017-02-01T11:23:12.000Z\\n"
"Content-Type: text/plain; charset=UTF-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"MIME-Version: 1.0\\n"
"X-Generator: react-intl-po\\n"


#: ./test/messages/src/containers/App/App.json
#. [App.Creator 1] - Creator 1
#. defaultMessage is:
#. Creator
msgid "App.Creator 1"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.Creator 2] - Creator 2
#. defaultMessage is:
#. Creator
msgid "App.Creator 2"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgid "App.errorButton"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorMessage] - The error message when api response as 404 not found
#. defaultMessage is:
#. The device is now private or deleted.
msgid "App.errorMessage"
msgstr ""

#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.Creator] - Creator
#. defaultMessage is:
#. Creator
msgid "NotFound.Creator"
msgstr ""

#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgid "NotFound.errorButton"
msgstr ""

`;

exports[`should return messages object with default mapper 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2017-02-01T11:23:12.000Z\\n"
"Content-Type: text/plain; charset=UTF-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"MIME-Version: 1.0\\n"
"X-Generator: react-intl-po\\n"


#: ./test/messages/src/containers/App/App.json
#. [App.Creator 1] - Creator 1
#. defaultMessage is:
#. Creator
#: ./test/messages/src/containers/App/App.json
#. [App.Creator 2] - Creator 2
#. defaultMessage is:
#. Creator
#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.Creator] - Creator
#. defaultMessage is:
#. Creator
msgid "Creator"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgid "Go to MCS website"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorMessage] - The error message when api response as 404 not found
#. defaultMessage is:
#. The device is now private or deleted.
msgid "The device is now private or deleted."
msgstr ""

`;

exports[`should return messages object with messageContext 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2017-02-01T11:23:12.000Z\\n"
"Content-Type: text/plain; charset=UTF-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"MIME-Version: 1.0\\n"
"X-Generator: react-intl-po\\n"


#: ./test/messages/src/containers/App/App.json
#. [App.Creator 1] - Creator 1
#. defaultMessage is:
#. Creator
msgctxt "App.Creator 1"
msgid "Creator"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.Creator 2] - Creator 2
#. defaultMessage is:
#. Creator
msgctxt "App.Creator 2"
msgid "Creator"
msgstr ""

#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.Creator] - Creator
#. defaultMessage is:
#. Creator
msgctxt "NotFound.Creator"
msgid "Creator"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgctxt "App.errorButton"
msgid "Go to MCS website"
msgstr ""

#: ./test/messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgctxt "NotFound.errorButton"
msgid "Go to MCS website"
msgstr ""

#: ./test/messages/src/containers/App/App.json
#. [App.errorMessage] - The error message when api response as 404 not found
#. defaultMessage is:
#. The device is now private or deleted.
msgctxt "App.errorMessage"
msgid "The device is now private or deleted."
msgstr ""

`;
